src/components/include/sl.h

```diff
@@ -438,22 +439,41 @@ sl_cs_exit_schedule_nospin_arg(struct sl_thd *to)
 
 	if (t->properties & SL_THD_PROPERTY_OWN_TCAP) {
 		assert(t->budget && t->period);
+		assert(sl_thd_tcap(t) != sl__globals()->sched_tcap);
 
 		if (t->last_replenish == 0 || t->last_replenish + t->period <= now) {
-			tcap_res_t currbudget;
+			tcap_res_t currbudget = 0;
+			cycles_t replenish    = now - ((now - t->last_replenish) % t->period);  
 
-			t->last_replenish = now;
-			currbudget        = (tcap_res_t)cos_introspect(ci, sl_thd_tcap(t), TCAP_GET_BUDGET);
-			/* TODO: need to change logic for SNDCAP with tcap_delegate, and error handling */
-			if (currbudget < t->budget && cos_tcap_transfer(sl_thd_rcvcap(t), sl__globals()->sched_tcap, (t->budget - currbudget), t->prio)) assert(0);
+			ret = 0;
+			if (likely(t->last_replenish)) currbudget = (tcap_res_t)cos_introspect(ci, sl_thd_tcap(t), TCAP_GET_BUDGET);
```

Comment:
doesn't matter what `last_replenish` is?  Just that it is set to some value?

Reply:
`last_replenish` is checked for expiry before entering this block.

I think it could be an unnecessary optimization because, `last_replenish` is `== 0` only the first time. So, I optimized to not `introspect` the first time.